docs/proposals/secure-registry.md

```diff
@@ -0,0 +1,75 @@
+# Secure Registry Support
+
+Table of contents
+- [Problem Statement](#problem-statement)
+- [Terminology](#terminology)
+- [Proposed Design](#proposed-design)
+- [Related Issues](#related-issues)
+
+## Problem Statement
+Currently odo only supports registry that is hosted by the platform that has publicly signed certificate, we should support secure registry so that user is able to store the confidential devfile to the registry and let the platform with certificate in user's trust store host the registry, authentication is needed on user side to access the platform.
+
+## Terminology
+Registry: registry is the place that stores index file (index.json) and devfile (devfile.yaml) so that user can catalog and create devfile component from the registry. The registry itself can be hosted on GitHub (GitHub-hosted registry) or Cluster (Cluster-hosted registry)
+
+Authentication method (Credential):
+- Username/Password: usually user has full access to the resource with Username/Password authentication method.
+- Personal Access Token (PAT): this is the recommended authentication method as user can grant limited resource access to the personal access token to make the resource access more secure. 
+
+## Proposed Design
+Support Scenarios:
+
+In order to make secure registry support feature more clear and specific, we should support the following scenarios:
+1. GitHub-hosted registry
+   1. GitHub public:
+   - Clients authenticate with GitHub personal access token
+   - TLS achieved with Git public CA signed certification in client trust store
+   2. GitHub Enterprise:
+   - Clients authenticate with GHE personal access token
+   - TLS achieved with GHE public CA signed certification in client trust store
+2. Cluster-hosted registry
+   - Clients authenticate with service account token
+   - TLS achieved with ingress gateway CA signed certification in client trust store
+
+Context:
+1. Given GitHub is going to depreciate basic authentication with username/password https://developer.github.com/changes/2020-02-14-deprecating-password-auth/, we have to only support personal access token authentication method for GitHub-hosted registry scenario.
+2. For cluster-hosted registry, the registry architecture would be creating a NGINX server to host the secure registry, then create a ingress gateway for the NGNIX server to let client access.
+
+Work flow to access secure registry:
+1. Collect Credential
+
+    Currently we have `odo registry add <registry name> <registry URL>` and `odo registry update <registry name> <registry URL>` to add and update registry accordingly. Regarding the CLI design, we can implement the following CLI design to support collecting credential:
+    - `odo registry add <registry name> <registry URL> --token <token>`
```

Comment:
What would be a command for adding registry with basic-auth?

I can see two options:

`odo registry add <registry name> <registry URL> --user <user> --password <password>`
or 
`odo registry add <registry name> <registry URL> --basic-auth <user>:<password>`

Reply:
I like the first option as it aligns with `kubectl`, `docker`, `oc` and `odo`, I am gonna update the design proposal.